### PR TITLE
ESQL: Profile more timing information

### PR DESCRIPTION
**Pull request by nik9000**
This profiles additional timing information for each individual driver. To the results from `profile` it adds the start and stop time for each driver. That was already in the task status. To the profile and task status it also adds the number of times the driver slept and some more detailed history about a few of those times.

Explanation time! The compute engine splits work into some number of `Drivers` per node. Each `Driver` is a single threaded entity - it runs on a thread for a while then does one of three things:
1. Finishes
2. Goes async because one of it's `Operator`s has gone async
3. Yields the thread pool because it has run for too long

This PR measures the second two. At this point only three operators can go async:
* ENRICH
* Reading from an empty exchange
* Writing to a full exchange

We're quite interested the these sleeps at the moment because they think they may be slowing things down. Here's what it looks like when a driver goes async because it wants to read from an empty exchange:

```
... the rest of the profile ...
        "sleeps" : {
          "counts" : {
            "exchange empty" : 2
          },
          "first" : [
            {
              "reason" : "exchange empty",
              "sleep" : "2024-08-13T19:45:57.943Z",
              "sleep_millis" : 1723578357943,
              "wake" : "2024-08-13T19:45:58.159Z",
              "wake_millis" : 1723578358159
            },
            {
              "reason" : "exchange empty",
              "sleep" : "2024-08-13T19:45:58.164Z",
              "sleep_millis" : 1723578358164,
              "wake" : "2024-08-13T19:45:58.165Z",
              "wake_millis" : 1723578358165
            }
          ],
          "last": [same as above]
```

Every time the driver goes async we count it in the `counts` map - grouped by the reason the driver slept. We also record the sleep and wake times for the first and last ten times the driver sleeps. In this case it only slept twice, so the `first` and `last` ten times is the same array.

This should give us a good sense about why drivers sleep while using a limited amount of memory per driver.

### CODE REVIEW

File: x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/operator/Driver.java [hunk 1]
```diff
@@ -394,18 +408,23 @@ void onComplete(ActionListener<Void> listener) {
...
+        StringBuilder reason = new StringBuilder();
+        for (IsBlockedResult r : results) {
+            r.listener().addListener(oneOf);
+            if (reason.isEmpty() == false) {
+                reason.append(" OR ");
+            }
+            reason.append(r.reason());
         }
-        return oneOf;
+        return new IsBlockedResult(oneOf, reason.toString());
```
Comment (nik9000) [inline x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/operator/Driver.java:427]: This string is the key in a hash table but I think in practice we won't have that many of them so I think it's ok. Unless we make a bunch more async operators. Then maybe we'd prefer just taking the first one.

File: x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/operator/Driver.java [hunk 2]
```diff
@@ -449,17 +476,28 @@ public DriverProfile profile() {
...
                 startTime,
-                System.currentTimeMillis(),
+                now,
                 prev.cpuNanos() + extraCpuNanos,
                 prev.iterations() + extraIterations,
                 status,
                 statusOfCompletedOperators,
-                activeOperators.stream().map(op -> new DriverStatus.OperatorStatus(op.toString(), op.status())).toList()
+                activeOperators.stream().map(op -> new DriverStatus.OperatorStatus(op.toString(), op.status())).toList(),
+                sleeps
```
Comment (nik9000) [inline x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/operator/Driver.java:500]: My hope is that this isn't too expensive to do for every time we go async. I'm not worried about the case where we run for five minutes and update the status or run for a zillion iterations. This is tiny compared to that no matter how you slice it. But if we have an operator that goes async all the time then recording the sleep is non-trivial.

File: x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/operator/Driver.java [hunk 3]
```diff
@@ -170,35 +180,36 @@ public DriverContext driverContext() {
      * thread to do other work instead of blocking or busy-spinning on the blocked operator.
      */
     SubscribableListener<Void> run(TimeValue maxTime, int maxIterations, LongSupplier nowSupplier) {
+        updateStatus(0, 0, DriverStatus.Status.RUNNING, "driver running");
```
Comment (nik9000) [inline x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/operator/Driver.java:183]: this line breaks a bunch of tests. i'm going to see if i can remove it.

**Review by dnhatn — APPROVED** [review]
LGTM. Thanks Nik!

File: x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/operator/DriverSleeps.java [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+     */
+    public record Sleep(String reason, long sleep, long wake) implements Writeable, ToXContentObject {
+        Sleep(StreamInput in) throws IOException {
+            this(in.readString(), in.readLong(), in.readLong());
+        }
+
+        @Override
+        public void writeTo(StreamOutput out) throws IOException {
+            out.writeString(reason);
+            out.writeLong(sleep);
```
[the remark this answers is not in the document]
Comment (nik9000) [reply x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/operator/DriverSleeps.java:46]: We've been writing timestamps as longs other places. I'll have think.

**Pull request merged**